src/scheduler/scheduler.rs

```diff
@@ -386,31 +391,38 @@ impl<VM: VMBinding> GCWorkScheduler<VM> {
             if all_parked {
                 // If there're any designated work, resume the workers and process them
                 if self.worker_group.has_designated_work() {
-                    self.worker_group.dec_parked_workers();
+                    assert!(
```

Comment:
Is this assertion correct? Between the check `poll_schedulable_work()` and this line, there could be work pushed to its `designated_work`.

Reply:
Currently, yes; but in general, maybe no.

Currently, only ordinary (non-coordinate) work packets can add designated work.  Because the entire `poll_slow` function is protected by a lock, if a GC worker reaches this point, all other GC workers must have already been blocked at the `wait(guard)` call below in this function (therefore cannot push any designated work), and the current GC worker must have checked if itself has any pending designated work at the `self.poll_schedulable_work` call above in this function.  So the current GC worker must not have any designated work at this point.

If we add another coordinator work packet (in addition to `StopMutators` which is the only work packet executed on the coordinator at this moment) in the future that can send designated work to GC workers, we will need to remove this assertion, and probably update the control flow here, too.

Reply:
Does it mean we can push work to designated work from a normal worker (and normal work packets), but we cannot push work to designated work from the coordinator (and coordinator work packets)? If this is the case, it sounds fine -- we just need to make it clear that a coordinator work packet should not push work to `designated_work`.

Reply:
That's right.  If we push work to designated work from coordinator, this assertion will fail. But a coordinator work packet has no problem adding more coordinator work packets, because it just pushes to the coordinator's message queue.

Reply:
Can you add some comments for this? For example, in `designated_work`, make it clear that we should only push to work to it from normal workers. For this assertion, make it clear that the assertion failure may be due to pushing work to `designated_work` from the coordinator.

Reply:
Yes.  I updated the comment for `designated_work` and mentioned that designated work cannot be added from the GC controller thread.